include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+template<
+    uint16_t GroupSize,
+    uint16_t KeyBucketCount,
+    typename Key,
+    typename KeyAccessor,
+    typename ValueAccessor,
+    typename ScratchAccessor,
+    typename SharedAccessor,
+    bool robust=false
+>
+struct counting
+{
+    uint32_t inclusive_scan(uint32_t value, NBL_REF_ARG(SharedAccessor) sdata)
+    {
+        return workgroup::inclusive_scan < plus < uint32_t >, GroupSize >::
+                template __call <SharedAccessor>(value, sdata);
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
+        uint32_t baseIndex = (glsl::gl_WorkGroupID().x * GroupSize) * data.elementsPerWT;
+
+        for (int i = 0; i < buckets_per_thread; i++) {
+            uint32_t prev_bucket_count = GroupSize * i;
+            sdata.set(prev_bucket_count + tid, 0);
+        }
```

Comment:
much easier (clearer and probably more performant) way to write this, see https://github.com/Devsh-Graphics-Programming/Nabla/pull/690/files#r1616056936
```cpp
for (uint32_t vid=tid; vid<KeyBucketCount; vid++)
    sdata.set(vid,0);
```

Reply:
also move the `baseIndex` calc till after the loop, it makes reading confusing

Reply:
Done